### PR TITLE
Resiliency Test: Reboot Node During PDS App Version Update

### PR DESCRIPTION
**Pull request by Madanagopal19**
1. Trigger app version upgrade through PDS
2. While app upgrade is happening, reboot one of the nodes thats hosting the app
3. Once app upgrade completes, trigger a workload and verify data

logs: https://aetos.pwx.purestorage.com/resultSet/testSetID/164643

### CODE REVIEW

File: tests/pds/dataservice_test.go [hunk 1]
```diff
@@ -1238,67 +1251,73 @@ func UpgradeDataService(dataservice, oldVersion, oldImage, dsVersion, dsBuild st
...
+		log.Infof("Version/Build: %v %v", dsVersion, dsBuild)
+		updatedDeployment, err := pdslib.UpdateDataServiceVerison(deployment.GetDataServiceId(), deployment.GetId(),
+			dataServiceDefaultAppConfigID,
+			replicas, dataServiceDefaultResourceTemplateID, dsBuild, dsVersion)
+		log.FailOnError(err, "Error while updating data services")
+		log.InfoD("data service deployed %v ", dataservice)
 
-			resourceTemp, storageOp, config, err := pdslib.ValidateDataServiceVolumes(updatedDeployment, dataservice, dataServiceDefaultResourceTemplateID, storageTemplateID, namespace)
-			log.FailOnError(err, "error on ValidateDataServiceVolumes method")
+		err = pdslib.ValidateDataServiceDeployment(updatedDeployment, namespace)
```
Comment (dbhatnagar-px) [inline tests/pds/dataservice_test.go:1292]: For now it's ok, but it's becoming a big mess soon. We should not be incorporating logic in test case file. Let's soon re-design our test case writing to just have a absolutely neccessary number of calls to functions that does all the magic and make it extremely modular. Plus let's remove all functions from test case file as well. 
@gauraa2, @Madanagopal19  & @jkagliwal-px : Let's sit in a meeting and re-design the tests. This model will cause our test case files to go beyond tens of thousands of lines soon - making the tests totally unmanageable. 

^ This is a generic comment not specific to this change. For now it is ok, but we will change it soon.

File: tests/pds/dataservice_test.go [hunk 2]
```diff
@@ -1157,8 +1157,21 @@ var _ = Describe("{DeployAllDataServices}", func() {
 	})
 })
 
-func TriggerDeployDataService(ds PDSDataService, namespace, tenantID, projectID string) (*pds.ModelsDeployment, map[string][]string, map[string][]string, error) {
+func TriggerDeployDataService(ds PDSDataService, namespace, tenantID, projectID string, deployOldVersion bool) (*pds.ModelsDeployment, map[string][]string, map[string][]string, error) {
```
Comment (dbhatnagar-px) [inline tests/pds/dataservice_test.go:1160]: Function UpgradeDataService is doing a similar task. Let's try to reuse either that or make changes in that to re-use this code.

File: tests/pds/resiliency_test.go [hunk 3]
```diff
@@ -165,6 +165,131 @@ var _ = Describe("{RebootActiveNodeDuringDeployment}", func() {
...
+				failuretype := pdslib.TypeOfFailure{
+					Type: ActiveNodeRebootDuringDeployment,
+					Method: func() error {
+						return pdslib.RebootActiveNodeDuringDeployment(params.InfraToTest.Namespace)
+					},
+				}
+				pdslib.DefineFailureType(failuretype)
+
+				log.Infof("Version/Build: %v %v", ds.Version, ds.Image)
+				updatedDeployment, err := pdslib.UpdateDataServiceVerison(deployment.GetDataServiceId(), deployment.GetId(),
```
Comment (dbhatnagar-px) [inline tests/pds/resiliency_test.go:220]: Won't this wait till Update is successful due to implicit call to "ValidateDataServiceDeployment" ? Your use case is to come out after just triggering update. Am I wrong ?

File: tests/pds/resiliency_test.go [hunk 4]
```diff
@@ -165,6 +165,131 @@ var _ = Describe("{RebootActiveNodeDuringDeployment}", func() {
...
+					dataServiceDefaultAppConfigID,
+					int32(ds.Replicas), dataServiceDefaultResourceTemplateID, ds.Image, ds.Version)
+				log.FailOnError(err, "Error while updating data services")
+				log.InfoD("data service %v update triggered", ds.Name)
+
+				err = pdslib.InduceFailureAfterWaitingForCondition(updatedDeployment, namespace, params.ResiliencyTest.CheckTillReplica)
```
Comment (dbhatnagar-px) [inline tests/pds/resiliency_test.go:226]: Issue with this line is , GetPdsSs will always return true immediately in this case without validating that upgrade has triggered or not - as there will be more than 1 pod active here for any multi-node deployment.

**Review by dbhatnagar-px — CHANGES_REQUESTED** [review]

File: tests/pds/dataservice_test.go (hunk 1 above)
Comment (Madanagopal19) [reply tests/pds/dataservice_test.go:1292]: sure

File: tests/pds/dataservice_test.go (hunk 2 above)
Comment (Madanagopal19) [reply tests/pds/dataservice_test.go:1160]: yes I have checked it, but UpgradeDataService function is having lot of validations which is not required and the easiest way of implementation is with this method.

File: tests/pds/resiliency_test.go (hunk 3 above)
Comment (Madanagopal19) [reply tests/pds/resiliency_test.go:220]: I have removed the `ValidateDataServiceDeployment ` from `UpdateDataServiceVerison ` method so we are good here.

File: tests/pds/resiliency_test.go (hunk 4 above)
Comment (Madanagopal19) [reply tests/pds/resiliency_test.go:226]: I have triggered the `UpdateDataServiceVerison ` method just before `InduceFailureAfterWaitingForCondition ` method, and that method handles the issue you mentioned `validating that upgrade has triggered or not`. For me there just be atleast 1 active pods to trigger the reboot.

File: tests/pds/dataservice_test.go (hunk 2 above)
Comment (dbhatnagar-px) [reply tests/pds/dataservice_test.go:1160]: Ok, in that case let's reuse this code in UpgradeDataService function.

File: tests/pds/resiliency_test.go (hunk 3 above)
Comment (dbhatnagar-px) [reply tests/pds/resiliency_test.go:220]: Exactly what we don't want as you moved the Validation part to test case directly. But as discussed above, let's take it up later.

**Review by dbhatnagar-px — CHANGES_REQUESTED** [review]

File: tests/pds/dataservice_test.go (hunk 1 above)
Comment (jkagliwal-px) [reply tests/pds/dataservice_test.go:1292]: I had same concern ! Lets re-design.

File: tests/pds/dataservice_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Madanagopal19) [reply tests/pds/dataservice_test.go:1160]: Lets have a story for optimisation and figure it out what can be optimised. Lets discuss and create task for it.

File: tests/pds/dataservice_test.go [hunk 5]
```diff
@@ -1195,7 +1208,7 @@ func TriggerDeployDataService(ds PDSDataService, namespace, tenantID, projectID
 
 func DeployandValidateDataServices(ds PDSDataService, namespace, tenantID, projectID string) (*pds.ModelsDeployment, map[string][]string, map[string][]string, error) {
 	log.InfoD("Data Service Deployment Triggered")
-	deployment, dataServiceImageMap, dataServiceVersionBuildMap, err := TriggerDeployDataService(ds, namespace, tenantID, projectID)
+	deployment, dataServiceImageMap, dataServiceVersionBuildMap, err := TriggerDeployDataService(ds, namespace, tenantID, projectID, false)
```
[the remark this answers is not in the document]
Comment (Madanagopal19) [reply tests/pds/dataservice_test.go:1211]: Its updated in all affected places and those changes are part of this PR

**Review by dbhatnagar-px — APPROVED** [review]
Lgtm

**Review by aravindranath14 — APPROVED** [review]
Looks good

**Pull request merged**